### PR TITLE
Sidechannel slashing

### PR DESCRIPTION
**Pull request by mankenavenkatesh**
Task List
- [x] Uptime Slashing
- [x] DoubleSign Slashing
- [x] Jail Validator
- [x] UnJail Validator
- [x] UnJail check in Current Validator state
- [x] Tick Mechanism
- [x] Submit Tick to Rootchain
- [x] Slashing Simulation  
- [x] Slashing Params 
- [x] Rest queries to fetch signing, slashing info.
- [x] Uptime, Double slash percent
- [x] Rest & cli api's to send tick, tick-ack, unjail.
- [x] Bridge subscribe to begin block tick event.
- [x] Add nonce to tick and tick-ack

Read More --- https://www.notion.so/maticnetwork/Slashing-b31c8eedab7d41be95c095f00c5251a2

Testing - https://www.notion.so/maticnetwork/Slashing-b31c8eedab7d41be95c095f00c5251a2#f2f8208d2bb44e9b9b1b8e2b69a1be06

### CODE REVIEW

File: helper/query.go [hunk 1]
```diff
@@ -346,4 +346,4 @@ func FetchSideTxSigs(
 
 	// return
 	return sigs, nil
-}
+}
```
Comment (0xAshish) [inline helper/query.go:349]: newline

File: slashing/keeper.go [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+
+// Slashing Info api's
+
+// SlashInterim - Add slash amounts to a buffer and emit <slash-limit> event if exceeded
+func (k *Keeper) SlashInterim(ctx sdk.Context, valID hmTypes.ValidatorID, slashPercent sdk.Dec) uint64 {
+	if slashPercent.IsNegative() {
+		panic(fmt.Errorf("attempted to slash with a negative slash factor: %v", slashPercent))
+	}
+
+	validator, _ := k.sk.GetValidatorFromValID(ctx, valID)
```
Comment (jdkanani) [inline slashing/keeper.go:264]: No check for validator error?

File: slashing/keeper.go [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+	store := ctx.KVStore(k.storeKey)
+
+	// get validator iterator
+	iterator := sdk.KVStorePrefixIterator(store, types.BufferValSlashingInfoKey)
+	defer iterator.Close()
+
+	// loop through validators to get valid validators
+	for ; iterator.Valid(); iterator.Next() {
+		// unmarshall validator
+		slashingInfo, _ := hmTypes.UnmarshallValSlashingInfo(k.cdc, iterator.Value())
```
Comment (jdkanani) [inline slashing/keeper.go:440]: No check of error?

File: slashing/msg_test.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	t.Log(hmTypes.ZeroHeimdallHash)
+
+	t.Log(msg.Proposer)
+	t.Log(msg.SlashingInfoHash)
+
+	t.Log(msg.Proposer.String())
+	t.Log(msg.SlashingInfoHash.String())
+
+	t.Log(hex.EncodeToString(msg.GetSideSignBytes()))
+}
```
[the remark this answers is not in the document]
Comment (mankenavenkatesh) [reply slashing/msg_test.go:46]: Hahaha....i added to check SideSignBytes. 
Not a test case.  Yeah....will add more test cases

File: slashing/genesis.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		signingInfos[valID.String()] = info
+		localMissedBlocks := []types.MissedBlock{}
+
+		keeper.IterateValidatorMissedBlockBitArray(ctx, valID, func(index int64, missed bool) (stop bool) {
+			localMissedBlocks = append(localMissedBlocks, types.NewMissedBlock(index, missed))
+			return false
+		})
+		missedBlocks[valID.String()] = localMissedBlocks
+		return false
+	})
```
[the remark this answers is not in the document]
Comment (mankenavenkatesh) [reply slashing/genesis.go:61]: not required i think. 
we always query by valID.

File: staking/keeper.go [hunk 6]
```diff
@@ -430,3 +439,58 @@ func (k *Keeper) IterateStakingSequencesAndApplyFn(ctx sdk.Context, f func(seque
...
+	// get validator from state
+	validator, found := k.GetValidatorFromValID(ctx, valSlashingInfo.ID)
+	k.Logger(ctx).Debug("validator fetched", "validator", validator)
+	if !found {
+		k.Logger(ctx).Error("Unable to fetch valiator from store")
+		return errors.New("validator not found")
+	}
+
+	// calculate power after slash
+	updatedPower := validator.VotingPower - int64(valSlashingInfo.SlashedAmount)
```
[the remark this answers is not in the document]
Comment (mankenavenkatesh) [reply staking/keeper.go:461]: Yeah...good to have. adding it.
It should be always true though as we slash until jail limit is reached. 
jail limit is percent of ValPower

File: slashing/infractions.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+	validator, err := k.sk.GetValidatorInfo(ctx, signerAddress.Bytes())
+	if err != nil {
+		k.Logger(ctx).Error("validator info not found", "address", signerAddress)
+		return err
+	}
+
+	// fetch signing info
+	signInfo, found := k.GetValidatorSigningInfo(ctx, validator.ID)
+	if !found {
+		panic(fmt.Sprintf("Expected signing info for validator %s but not found", addr))
```
[the remark this answers is not in the document]
Comment (mankenavenkatesh) [reply slashing/infractions.go:28]: validator signing info is created on validator-join. so there cannot be validator without signing info

**Pull request merged**